### ISSUE TITLE
The unit of the variable for @for is inconsistent between implementations

### ISSUE THREAD

**Issue by stof** [sass/sass#2931]
With this input

```scss
test {
    @for $i from 12px to 13px {
        content: unit($i);
    }
}
```

I get the following output on ruby-sass 3.7.4 and libsass 3.5.5 (these are the versions available on https://www.sassmeister.com/):

```css
test {
  content: "px";
}
```

But I get this output on dart-sass 1.26.10 and 1.26.11

```css
test {
  content: "";
}
```

IMO, the behavior of ruby-sass and libsass is more logical, as the values for `from` and `to` are handled as unit numbers (including coercing compatible units).

**Comment by stof** [comment]
Looking at the behavior with different units, the behavior is even more inconsistent:

- dart-sass coerce to the unit of `to` (but then looses the unit for `$i`, making it hard to use as we might have weird numbers)
- ruby-sass coerce to the unit of `from`
- libsass does not coerce compatible units. It requires the same unit for `from` and `to`

Given that `from` and `to` are required to be integers *after* converting them to the common unit, it makes it impossible to write portable sass code when units are not the same.

**Comment by stof** [comment]
Looking at `@for` in sass-spec, it always uses unitless numbers for non-error tests. There is a single test using unit, which covers the case of incompatible units (`%` and `px`) but the case of compatible units is not tested.

**Comment by jathak** [comment]
Good catch! It seems like Ruby Sass's behavior here is probably the best option here, so we'll write up a [fast-track proposal](https://github.com/sass/sass/blob/master/CONTRIBUTING.md#fast-track) to formally specify this, and then update the behavior of Dart Sass.

**Comment by babakfp** [comment]
> logical

Just curious, why you used `px` (for loop parameters)?

**Comment by babakfp** [comment]
> With this input
> 
> ```css-scss
> test {
>     @for $i from 12px to 13px {
>         content: unit($i);
>     }
> }
> ```
> 
> I get the following output on ruby-sass 3.7.4 and libsass 3.5.5 (these are the versions available on https://www.sassmeister.com/):
> 
> ```css
> test {
>   content: "px";
> }
> ```
> 
> But I get this output on dart-sass 1.26.10 and 1.26.11
> 
> ```css
> test {
>   content: "";
> }
> ```
> 
> IMO, the behavior of ruby-sass and libsass is more logical, as the values for `from` and `to` are handled as unit numbers (including coercing compatible units).

Just curious, why you used `px` (for loop parameters)?

**Comment by babakfp** [comment]
Just curious, why you used `px` (for loop parameters)?

**Comment by stof** [comment]
@babakfp I don't have a use case myself. I was working on the handling of unit numbers in the scssphp library and I noticed that our handling of units for `@for` was not the same than in dart-sass (I was looking at the implementation directly). So I went into looking at how other implementations deal with that, and I discovered this mess.

Regarding the expected behavior, I think that the best behavior is either the ruby-sass or libsass one. If I define the loop with unit, the loop variable should should also have the unit. And then, coercing compatible units (like ruby-sass) would be more consistent with other parts of sass. But we need to  define which side `from` or `to` defines the unit being used, as that's not interoperable (due to the requirement to use integers, selecting a different unit might make the loop invalid).
